pkg/util/retry.go

```diff
@@ -41,3 +43,46 @@ func Retry(maxRetries int, delay time.Duration, f func() error) error {
 		<-time.After(delay)
 	}
 }
+
+// RetryFunc is a function that returns true when it is done and it should be retried no longer.
+// It should return error if there has been an error. The error will be logged if done==false
+// (should keep retrying). The error will be returned by the calling function if done==true (should
+// stop retrying).
+type RetryFunc func() (done bool, err error)
+
+// RetryWithTimeout retries the RetryFunc until the timeout occurs. It will retry once more after
+// the timeout to avoid race conditions at the expense of running for slightly longer than timeout
+// in the timeout/error case.
+// The given description will be output in log/error messages as "... waiting for <description>..."
+func RetryWithTimeout(f RetryFunc, period time.Duration, timeout time.Duration, description string) error {
+	tt := time.After(timeout)
+	var done bool
+	var err error
+	for {
+		done, err = f()
+		if done {
+			break
+		}
+		if err != nil {
+			logger.Errorf("error occurred waiting for %s. retrying after %v. %v", description, period, err)
+		}
+		logger.Debugf("waiting for %s. retrying after %v", description, period)
+
+		select {
+		case <-time.After(period):
```

Comment:
Should we have a cancellation context too? Or just check if the orch should be cancelled?

Reply:
This isn't currently called from a goroutine, so if I add a cancellation context, it won't ever be used.

Reply:
What about the check for cancellation then?

Reply:
Are you requesting to check for a cancelled reconcile from this function? I can do that, but I think that makes some of update logic bleed from the Ceph operator/controller down to the util package.

I don't anticipate this retry being more than 10 minutes in real world cases, which doesn't stall reconciliation for very long. 

The existing updatedeploymentandwait function doesn't look for cancellation. I don't think it's bad to add  a cancellation context here, but we also should add one to `updatedeploymentandwait` if we are going to do that, and then I think that's something for another PR.

TBH, I would like to clean up how the existing `UpdateDeploymentAndWait` is implemented. There are two callbacks. It seems unnecessarily complicated, and it means that some of the update logic that is unique to Ceph bleeds down into low-level helpers here. There could be a good opportunity to change all this at the same time.

Reply:
Sounds good for a future PR, the callbacks were made initially because other operators were using `UpdateDeploymentAndWait`.